ENIGMAsystem/SHELL/Universal_System/Extensions/ProcInfo/MacOSX/Cocoa/procinfo.mm

```diff
@@ -0,0 +1,162 @@
+/*
+
+ MIT License
+ 
+ Copyright © 2020 Samuel Venable
+ 
+ Permission is hereby granted, free of charge, to any person obtaining a copy
+ of this software and associated documentation files (the "Software"), to deal
+ in the Software without restriction, including without limitation the rights
+ to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ copies of the Software, and to permit persons to whom the Software is
+ furnished to do so, subject to the following conditions:
+ 
+ The above copyright notice and this permission notice shall be included in all
+ copies or substantial portions of the Software.
+ 
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ SOFTWARE.
+ 
+*/
+
+#import "subclass.h"
+#import <Cocoa/Cocoa.h>
+#import <sys/types.h>
+#import <unistd.h>
+
+NSWindow *cocoa_window_from_wid(CGWindowID wid) {
+  return [NSApp windowWithWindowNumber:wid];
+}
+
+CGWindowID cocoa_wid_from_window(NSWindow *window) {
+  return [window windowNumber];
+}
+
+bool cocoa_wid_exists(CGWindowID wid) {
+  bool result = false;
+  const CGWindowLevel kScreensaverWindowLevel = CGWindowLevelForKey(kCGScreenSaverWindowLevelKey);
+  CFArrayRef windowArray = CGWindowListCopyWindowInfo(kCGWindowListOptionAll, kCGNullWindowID);
+  CFIndex windowCount = 0;
+  if ((windowCount = CFArrayGetCount(windowArray))) {
+    for (CFIndex i = 0; i < windowCount; i++) {
+      NSDictionary *windowInfoDictionary =
+      (__bridge NSDictionary *)((CFDictionaryRef)CFArrayGetValueAtIndex(windowArray, i));
+      NSNumber *ownerPID = (NSNumber *)(windowInfoDictionary[(id)kCGWindowOwnerPID]);
+      NSNumber *level = (NSNumber *)(windowInfoDictionary[(id)kCGWindowLayer]);
+      if (level.integerValue < kScreensaverWindowLevel) {
+        NSNumber *windowID = windowInfoDictionary[(id)kCGWindowNumber];
+        if (wid == windowID.integerValue) {
+          result = true;
+          break;
+        }
+      }
+    }
+  }

```

Comment:
All of this logic is repeated quite a bit, it's another case for a helper function. Move the loop and all the general logic to a helper that accepts a callback function pointer or `std::function` and then call that inside the loop. Then all of the actual functions can call the helper function and pass a lambda for the logic.

This may seem kind of difficult if you aren't familiar with lambda's already, so I wrote some almost-pseudo code for you. This is for the `cocoa_pid_from_wid` function, but you can adapt the callback to accept more parameters or however many is necessary to implement the body logic of each of these functions. You'll have to fiddle with it a bit to get this to work probably.
```cpp
void loop_helper(std::function<void(NSNumber*)> callback) {
  const CGWindowLevel kScreensaverWindowLevel = CGWindowLevelForKey(kCGScreenSaverWindowLevelKey);
  CFArrayRef windowArray = CGWindowListCopyWindowInfo(kCGWindowListOptionAll, kCGNullWindowID);
  CFIndex windowCount = 0;
  if ((windowCount = CFArrayGetCount(windowArray))) {
    for (CFIndex i = 0; i < windowCount; i++) {
      NSDictionary *windowInfoDictionary =
      (__bridge NSDictionary *)((CFDictionaryRef)CFArrayGetValueAtIndex(windowArray, i));
      NSNumber *ownerPID = (NSNumber *)(windowInfoDictionary[(id)kCGWindowOwnerPID]);
      NSNumber *level = (NSNumber *)(windowInfoDictionary[(id)kCGWindowLayer]);
      if (level.integerValue < kScreensaverWindowLevel) {
        NSNumber *windowID = windowInfoDictionary[(id)kCGWindowNumber];
        if (wid == windowID.integerValue) {
          callback(ownerPID);
          break;
        }
      }
    }
  }
  CFRelease(windowArray);
}

pid_t cocoa_pid_from_wid(CGWindowID wid) {
  pid_t pid;
  loop_helper([&](NSNumber *ownerPID) {
    pid = ownerPID.integerValue;
  });
  return pid;
}
```

Reply:
Thanks for the suggestion. Unfortunately I've never used C++ and Objective-C in the same file, (I guess in which case it would be considered Objective-C++) and when I use the compiler flag for it  which is -ObjC++ instead of -ObjC apparently there are some other things that need to be done to use C++ features in the same file, because I've been wondering how to do it for a while now with no success, I get unresolved symbols for some reason. Which is why I have been using C with it in various places instead. I'll have to look more into that because C++ would be a lot more desirable to use for me.

Anyway, I've tried making a helper function to cover all this in the past, but it made the code so very unreadable that I messed up in several places and had to revert the code because I made it no longer work by accident. Objective-C is needlessly verbose. 

I'm tempted to buy that Objective-C to Swift code converter so we have something that is more modern, so it'll last longer, and be much more brief in length. Anyway, thanks!

Reply:
That was my mistake, I wasn't thinking about objective C at all. Do you have a similar kind of loop on the other platforms? If so please try what I wrote on Linux/Windows first. We can work through making an objective C version later, it shouldn't be too difficult to find a replacement for `std::function` I don't think. Is it difficult for you to translate what I did into Windows/Linux C++? If so, I can make an attempt to rewrite it for those platforms.

Reply:
Don't worry about it, I'll try to make a helper function again for this.